src/pattern_matching/myers.rs

```diff
@@ -13,75 +13,246 @@
+    }
+}
+
+/// Myers instance based on 32-bit integers (pattern length up to 32)
+pub type Myers32 = Myers<u32>;
+/// Myers instance based on 64-bit integers (pattern length up to 64)
+pub type Myers64 = Myers<u64>;
+/// Myers instance based on 128-bit integers (pattern length up to 128)
+#[cfg(has_u128)]
+pub type Myers128 = Myers<u128>;
+
 
 /// Myers algorithm.
-pub struct Myers {
-    peq: [u64; 256],
-    bound: u64,
-    m: u8,
+pub struct Myers<T = u64>
+where T: Num
+{
+    peq: [T; 256],
+    bound: T,
+    m: usize,
+    tb: Traceback<T>,
 }
 
-impl Myers {
+
+
+impl<T: Num> Myers<T> {
     /// Create a new instance of Myers algorithm for a given pattern.
-    pub fn new<'a, P: IntoTextIterator<'a>>(pattern: P) -> Self
-    where
-        P::IntoIter: ExactSizeIterator,
+    pub fn new<'a, P: IntoTextIterator<'a>>(pattern: P) -> Self {
+        Self::with_ambig(pattern.into_iter().map(Some))
+    }
+
+    // TODO: eventually move to module docs and use shorter example
+    /// Like `Myers64::new()`, but additionally allows for specifying
+    /// multiple matching characters in a pattern.
+    /// As in `new`, an iterator over pattern positions is supplied. However, a nested iterable
+    /// additionally allows to return multiple characters which should be recognized
+    /// as matching at a given position. Unambiguous positions can be specified as iterables
+    /// with a single element (see below). Even positions that don't match any character are
+    /// possible by specifying an empty iterable.
+    ///
+    /// *Note*: Only a single indirection is possible, thus patterns with 'double borrows' like
+    ///  this one: `&[&b"abc"[..], &b"de"[..]]` are not possible. The outer slice could be replaced
+    /// by a `vec!` to make it work.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// # extern crate bio;
+    /// use bio::pattern_matching::myers::Myers64;
+    ///
+    /// # fn main() {
+    /// let text =    b"TGACGNTGA";
+    /// let pattern = b"TGANGCTGA";
+    ///
+    /// // 'N' has no special meaning:
+    /// let myers = Myers64::new(pattern);
+    /// assert_eq!(myers.distance(text), 2);
+    ///
+    /// // 'N' in a pattern matches all four bases, but 'N' in
+    /// // the text does not match any (asymmetric matching):
+    /// let myers_ambig_asymm = Myers64::with_ambig(pattern.into_iter().map(|&b| {
+    ///     // replacing N with all possible bases.
+    ///     // Using vectors. Otherwise, the ref_slice crate could be used
+    ///     if b == b'N' { b"ACGT".to_vec() } else { vec![b] }
+    /// }));
+    /// assert_eq!(myers_ambig_asymm.distance(text), 1);
+    ///
+    /// // 'N' matches both ways:
+    /// let myers_ambig_symm = Myers64::with_ambig(pattern.into_iter().map(|&b| {
+    ///     if b == b'N' { b"ACGT".to_vec() } else { vec![b, b'N'] }
+    /// }));
+    /// assert_eq!(myers_ambig_symm.distance(text), 0);
+    /// # }
+    pub fn with_ambig<P, I, C>(pattern: P) -> Self
```

Comment:
I think it could be easier to handle this with a builder pattern (best using the [builder crate](https://docs.rs/derive_builder/0.5.1/derive_builder/)). Something like

```
Myers::new(pattern).ambig(b'N', b"ACGT").build();
```

Reply:
Ah yes, that's a good idea, seems much simpler. However, I'm not sure how to implement this. The pattern is not stored directly, but 'encoded' in the `peq` array. I guess it would need to be stored in a separate `Vec`. Still, after each call to `.ambig()`, we would have to search the pattern for occurrences of this character. 

I guess in most cases the negative performance impact is neglectable since the `Myers` instance may be constructed once and used to search many sequences. Still, I don't know.


One possibility for simplification of my current implementation came into my mind: How about a trait similar to [std::str::pattern](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html), which would allow to specify either an `IntoIterator<Item=u8>`, but also nested iterators. I'm not sure whether that could easily be implemented, I would have to try first. Of course, it's still not as ergonomic as the builder pattern.

Reply:
I think I still prefer the builder. The builder struct itself can store each ambig declaration in a Vec. Then, upon build you can parse the pattern and perform the replacement. This way, multiple calls to ambig do not create the kind of overhead you describe, because the pattern only has to be converted once.

Reply:
Now I get it, yes that sounds very good. Maybe, this `MyersBuilder` (or how it would be called) could still be optional (just for the case of adding ambiguities) and the `Myers::new` method could be preserved.

Reply:
Yes, I agree.